### PR TITLE
Engine Message Queue Prototypes

### PR DESCRIPTION
**Pull request by jordanschalm**
2 draft solutions for https://github.com/onflow/flow/issues/396.

## 1 - Poker

Package: `mqueue/poker`

This is similar to the example code shared during the meeting on Tuesday. It demonstrates the minimal communication requirements between the `network->queue` step and the `queue->engine` step of the processing pipeline to allow responsive processing of messages. This does not restrict engines to channel-based message selection -- they may select messages from their inbound message queues however they like. The downside is that the queues must be thread-safe. 

## 2 - Real-time Queue

Package: `mqueue/rtqueue`

This is based off @AlexHentschel's suggestion to move the "shovelling" of messages between the two steps in the processing pipeline into the queue. This maintains the design where the inbound (`queue->engine`) message queues enforce a channel-based message selection so the underlying queue structure doesn't need to be thread-safe. Downside is message prioritization becomes more difficult. This implementation notably only uses a goroutine when the queue is full, and handles terminating the goroutine when the queue is empty. The benefit is we don't need to make the `Queue` ready-done-aware to gracefully stop its spawned goroutine, we just need to drain the queue when exiting the engine.

### CODE REVIEW

File: mqueue/poker/engine.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	// calling ProcessMessageFromNetwork
+	messages *fifoqueue.FifoQueue
+	poker    Poker
+	unit     *engine.Unit
+}
+
+// the network->queue step processor just adds the item to the queue.
+func (e *Engine) ProcessMessageFromNetwork(message mqueue.Message) {
+	e.messages.Push(message)
+}
```
Comment (jordanschalm) [inline mqueue/poker/engine.go:21]:
```suggestion
	e.messages.Push(message)
	e.poker.Poke()
}
```

File: mqueue/rtqueue/rtqueue.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+		shovelling: atomic.NewBool(false),
+	}
+	return q
+}
+
+// shovel is a goroutine that lives for as long as the queue is non-empty.
+func (q *queue) shovel() {
+
+	// shovel is always started when at least one item is being added
+	head := <-q.in
```
Comment (AlexHentschel) [inline mqueue/rtqueue/rtqueue.go:40]: Sorry, I might be completely misunderstanding. Here my thoughts:
* you are directly reading from the tail channel here
* so if you have elements in the `items`: won't this tail element will jump the queue (?) https://github.com/onflow/flow-go/blob/2f8380cdae6c1f709179b24766eac884e5b6eb7c/mqueue/rtqueue/rtqueue.go#L48
* I think you would need to first check `items`, grep the head from `items` if there is any. Only if `items` is empty, you can read from `q.in`

File: mqueue/rtqueue/rtqueue.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+		select {
+		case tail := <-q.in:
+			// when a new item is added, re-enter this select with the same head
+			q.items.Push(tail)
+			continue
+		case q.out <- head:
+			// when the head of the queue is accepted, retrieve the next
+			// head before re-entering this select
+		}
+
```
Comment (AlexHentschel) [inline mqueue/rtqueue/rtqueue.go:52]: I think an element could get stuck:
* lets assume you have just pushed the last element into `q.out`: https://github.com/onflow/flow-go/blob/2f8380cdae6c1f709179b24766eac884e5b6eb7c/mqueue/rtqueue/rtqueue.go#L48
* now, `q.items` is empty, so `ok` is `false`
* but before you get to https://github.com/onflow/flow-go/blob/2f8380cdae6c1f709179b24766eac884e5b6eb7c/mqueue/rtqueue/rtqueue.go#L56, some thread pushes another element into `q.in`. 
    * The current go routine won't notice it and subsequently exit
    * but at the time when https://github.com/onflow/flow-go/blob/2f8380cdae6c1f709179b24766eac884e5b6eb7c/mqueue/rtqueue/rtqueue.go#L64-L66 is evaluated, `q.shovelling` is still `false` so no new routine will be spawned to process the element ... 
   
   The culprit here is the non-atomicity of https://github.com/onflow/flow-go/blob/2f8380cdae6c1f709179b24766eac884e5b6eb7c/mqueue/rtqueue/rtqueue.go#L53-L58 and  https://github.com/onflow/flow-go/blob/2f8380cdae6c1f709179b24766eac884e5b6eb7c/mqueue/rtqueue/rtqueue.go#L64-L67 
     - you can circumvent this problem by not letting the go routine exit; then you arrive at solution https://github.com/onflow/flow-go/pull/460
     - alternatively, you can make these operations atomic by using a mutex. But then you don't need the `q.in` channel anymore and the thread executing https://github.com/onflow/flow-go/blob/2f8380cdae6c1f709179b24766eac884e5b6eb7c/mqueue/rtqueue/rtqueue.go#L63 could directly write to into the FifoQueue. The outbound channel is then a matter of taste: do I rather use a case statement to block? on channels or to I use the poker and poll the channels?

**Review by AlexHentschel — COMMENTED** [review]
Thanks for drafting this @jordanschalm . I like the Poker-based implementation. Regarding the real-time queue one, I have strong concerns regarding concurrency safety and messages getting stuck -- but that might be because I don't really understand.

**Pull request closed without merging**